widget/table.go

```diff
@@ -15,50 +15,41 @@ const tableDividerThickness = 1
 // Declare conformity with Widget interface.
 var _ fyne.Widget = (*Table)(nil)
 
+// TableCellID is a type that represents a cell's position in a table based on it's row and column location.
+type TableCellID struct {
+	Row int
+	Col int
+}
+
 // Table widget is a grid of items that can be scrolled and a cell selected.
 // It's performance is provided by caching cell templates created with CreateCell and re-using them with UpdateCell.
 // The size of the content rows/columns is returned by the Length callback.
 type Table struct {
 	BaseWidget
 
-	Length     func() (int, int)
-	CreateCell func() fyne.CanvasObject
-	UpdateCell func(row, col int, template fyne.CanvasObject)
-	// OnSelectionChanged is called whenever the selection changes, such as through user tap.
-	// The row and col values will be -1 if no cell is currently selected.
-	OnSelectionChanged func(row, col int)
-
-	selectedRow, selectedColumn int
-	hoveredRow, hoveredColumn   int
-	cells                       *tableCells
-	moveCallback                func()
-	offset                      fyne.Position
-	scroll                      *ScrollContainer
+	Length       func() (int, int)
+	CreateCell   func() fyne.CanvasObject
+	UpdateCell   func(id TableCellID, template fyne.CanvasObject)
+	OnSelected   func(id TableCellID)
+	OnUnselected func(id TableCellID)
+
+	selectedCell, hoveredCell *TableCellID
+	cells                     *tableCells
+	moveCallback              func()
+	offset                    fyne.Position
+	scroll                    *ScrollContainer
 }
 
 // NewTable returns a new performant table widget defined by the passed functions.
 // The first returns the data size in rows and columns, second parameter is a function that returns cell
 // template objects that can be cached and the third is used to apply data at specified data location to the
 // passed template CanvasObject.
-func NewTable(length func() (int, int), create func() fyne.CanvasObject, update func(int, int, fyne.CanvasObject)) *Table {
-	t := &Table{Length: length, CreateCell: create, UpdateCell: update, selectedRow: -1, selectedColumn: -1}
+func NewTable(length func() (int, int), create func() fyne.CanvasObject, update func(TableCellID, fyne.CanvasObject)) *Table {
+	t := &Table{Length: length, CreateCell: create, UpdateCell: update, selectedCell: nil, hoveredCell: nil}
```

Comment:
nit: selectedCell and hoveredCell will already be nil, no need to explicitly assign them to nil

Reply:
Resolved